Error if function pointer is declared in a safe context. Support local function pointer declaration.

333fred opened the pull request:
@AlekseyTs @dotnet/roslyn-compiler for a quick review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -19,13 +19,20 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
+        {
+            if (!typeBinder.InUnsafeRegion)
+            {
+                diagnostics.Add(ErrorCode.ERR_UnsafeNeeded, syntax.Location);
```
>syntax.Location [](start = 60, length = 15)

Consider reducing the span for the error. Perhaps to the ```delegate*```.

333fred replied:
Why? Personally, I'd prefer it to be on the whole thing.

---
In reply to: [376031475](https://github.com/dotnet/roslyn/pull/41455#discussion_r376031475) [](ancestors = 376031475)

AlekseyTs replied:
> Why? Personally, I'd prefer it to be on the whole thing.

What is the benefit of squiggling the whole signature, which, given the nature of the type, could span multiple lines? The benefit of reducing the span is to avoid intersection with other errors in the type, which makes it easier to locate their source.

---
In reply to: [376094583](https://github.com/dotnet/roslyn/pull/41455#discussion_r376094583) [](ancestors = 376094583,376031475)

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -19,13 +19,20 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
+            if (!typeBinder.InUnsafeRegion)
+            {
+                diagnostics.Add(ErrorCode.ERR_UnsafeNeeded, syntax.Location);
+            }
```
It feels like this should be reported in the caller.

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs:
```diff
@@ -417,6 +417,13 @@ internal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<Arra
                     var pointerTypeSyntax = (PointerTypeSyntax)type;
                     type = pointerTypeSyntax.ElementType;
                     goto recurse;
+                case SyntaxKind.FunctionPointerType:
```
>case SyntaxKind.FunctionPointerType [](start = 16, length = 35)

Did you check that other places that use ```SyntaxKind.PointerType``` are properly adjusted?

AlekseyTs replied:
>case SyntaxKind.FunctionPointerType: [](start = 16, length = 36)

Do we have a test that makes presence of this ```case``` observable?

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [583070113](https://github.com/dotnet/roslyn/pull/41455#issuecomment-583070113) [](ancestors = 583070113)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -435,9 +436,21 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
+                                     asterisk.LeadingWidth + asterisk.Width;
+
+                        RoslynDebug.Assert(@delegate.SyntaxTree is object);
+                        diagnostics.Add(info, Location.Create(@delegate.SyntaxTree, new TextSpan(@delegate.SpanStart, length)));
```
>new TextSpan(@delegate.SpanStart, length [](start = 84, length = 40)

```FromBounds(@delegate.SpanStart, asterisk.SpanEnd)```?

333fred replied:
Thanks, I looked for a helper and apparently missed it.

---
In reply to: [376166854](https://github.com/dotnet/roslyn/pull/41455#discussion_r376166854) [](ancestors = 376166854)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -20,12 +20,12 @@ internal sealed partial class FunctionPointerTypeSymbol : TypeSymbol
-                    diagnostics,
-                    basesBeingResolved,
-                    suppressUseSiteDiagnostics));
+                           FunctionPointerMethodSymbol.CreateFromSource(
```
>FunctionPointerMethodSymbol.CreateFromSource( [](start = 27, length = 45)

It looks like this file has no meaningful changes, consider reverting.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -374,9 +374,10 @@ internal override bool HasPointerType
                     Debug.Assert(_lazyType.Value.DefaultType.IsPointerType() ==
-                        IsPointerFieldSyntactically());
+                        IsPointerFieldSyntactically() ||
+                        _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer);
```
>_lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer [](start = 24, length = 62)

The purpose of the assert was to make sure result ```sPointerFieldSyntactically``` is consistent with result of binding. I think the change leaves this purpose unserved.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -374,9 +374,10 @@ internal override bool HasPointerType
                 if (_lazyType != null)
                 {
                     Debug.Assert(_lazyType.Value.DefaultType.IsPointerType() ==
```
>_lazyType.Value.DefaultType [](start = 33, length = 27)

Consider extracting a local.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -374,9 +374,10 @@ internal override bool HasPointerType
+                        _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer);
 
-                    return _lazyType.Value.DefaultType.IsPointerType();
+                    return _lazyType.Value.DefaultType.IsPointerType() || _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer;
```
>_lazyType.Value.DefaultType.IsPointerType() || _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer [](start = 27, length = 109)

Consider using switch expression over symbol kind, also putting the value into the local and using the local in the assert above (what we return here and the value we get from ```IsPointerFieldSyntactically``` should match).

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -386,7 +387,7 @@ internal override bool HasPointerType
         {
             var declaration = GetFieldDeclaration(VariableDeclaratorNode).Declaration;
-            if (declaration.Type.Kind() == SyntaxKind.PointerType)
+            if (declaration.Type.Kind() == SyntaxKind.PointerType || declaration.Type.Kind() == SyntaxKind.FunctionPointerType)
```
>declaration.Type.Kind() == SyntaxKind.PointerType || declaration.Type.Kind() == SyntaxKind.FunctionPointerType [](start = 16, length = 110)

switch expression?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:
```diff
@@ -559,7 +559,7 @@ internal bool HasPointerType
                 RefKind refKind;
                 var typeSyntax = syntax.Type.SkipRef(out refKind);
-                return typeSyntax.Kind() == SyntaxKind.PointerType;
+                return typeSyntax.Kind() == SyntaxKind.PointerType || typeSyntax.Kind() == SyntaxKind.FunctionPointerType;
```
>typeSyntax.Kind() == SyntaxKind.PointerType || typeSyntax.Kind() == SyntaxKind.FunctionPointerType [](start = 23, length = 98)

switch expression?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -374,9 +374,10 @@ internal override bool HasPointerType
+                        _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer);
 
-                    return _lazyType.Value.DefaultType.IsPointerType();
+                    return _lazyType.Value.DefaultType.IsPointerType() || _lazyType.Value.DefaultType.Kind == SymbolKind.FunctionPointer;
```
>IsPointerType [](start = 55, length = 13)

I assume you also checked other places where we check for regular pointer type, by using this helper or type/symbol kind.

333fred replied:
Yes. These were the only places I feel comfortable changing at this point. I did mean to add a prototype comment to `IsPointerType` to follow up on some more places, but I apparently missed staging it.

---
In reply to: [376482345](https://github.com/dotnet/roslyn/pull/41455#discussion_r376482345) [](ancestors = 376482345)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -692,5 +698,92 @@ .maxstack  8
+                    // (6,13): warning CS0219: The variable 'a' is assigned but its value is never used
+                    //         int a = 1;
+                    Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "a").WithArguments("a").WithLocation(6, 13),
+                    // (7,22): error CS0270: Array size cannot be specified in a variable declaration (try initializing with a 'new' expression)
```
>try initializing with a 'new' expression [](start = 103, length = 40)

It doesn't look like this portion is applicable to the situation, but it is probably preexisting condition.

333fred replied:
Do you think this is worth adding a new error for it? Personally, I can't see this particular scenario happening very often.

---
In reply to: [376485842](https://github.com/dotnet/roslyn/pull/41455#discussion_r376485842) [](ancestors = 376485842)

AlekseyTs replied:
> Do you think this is worth adding a new error for it?

It might be worth opening an issue, I assume there are scenarios not related to this feature that are also affected. Like within type arguments, etc.

---
In reply to: [376608344](https://github.com/dotnet/roslyn/pull/41455#discussion_r376608344) [](ancestors = 376608344,376485842)

333fred replied:
[https://github.com/dotnet/roslyn/issues/41494](https://github.com/dotnet/roslyn/issues/41494)

---
In reply to: [376609526](https://github.com/dotnet/roslyn/pull/41455#discussion_r376609526) [](ancestors = 376609526,376608344,376485842)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -692,5 +698,92 @@ .maxstack  8
+        }
+
+        [Fact]
+        public void MisdeclaredArraysWithLocalDeclarationsAreHandled()
```
>MisdeclaredArraysWithLocalDeclarationsAreHandled [](start = 20, length = 48)

It would be good to add tests covering other code paths you adjusted.

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [583487512](https://github.com/dotnet/roslyn/pull/41455#issuecomment-583487512) [](ancestors = 583487512)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -435,9 +436,21 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
+                    {
+                        var @delegate = functionPointerTypeSyntax.DelegateKeyword;
+                        var asterisk = functionPointerTypeSyntax.AsteriskToken;
+                        var length = @delegate.Width + @delegate.TrailingWidth +
```
>length [](start = 28, length = 6)

Remove?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:
```diff
@@ -1053,6 +1053,7 @@ public static bool IsPartial(this TypeSymbol type)
             return type is SourceNamedTypeSymbol { IsPartial: true };
         }
 
+        // PROTOTYPE(func-ptr): Follow up on usages of this API and add FunctionPointers where necessary.
```
>Follow up on usages of this API and add FunctionPointers where necessary. [](start = 32, length = 73)

Consider also mentioning symbol/type kind checks.
